File: app/src/main/java/com/fazemeright/chatbotmetcs622/ui/base/BaseActivity.java
package com.fazemeright.chatbotmetcs622.ui.base;

import android.app.Activity;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import androidx.annotation.LayoutRes;
import androidx.annotation.MenuRes;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.fazemeright.chatbotmetcs622.R;
import com.fazemeright.chatbotmetcs622.network.ApiManager;
import com.fazemeright.chatbotmetcs622.network.NetworkManager;
import com.fazemeright.chatbotmetcs622.repositories.MessageRepository;

public abstract class BaseActivity extends AppCompatActivity {

  public Context mContext;
  protected MessageRepository messageRepository;
  protected ApiManager apiManager;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    mContext = this;
    messageRepository = MessageRepository.getInstance(mContext);
    apiManager = ApiManager.getInstance();
    apiManager.init(NetworkManager.getInstance());
    setContentView(getLayoutResId());
  }

  @Override
  public void setContentView(@LayoutRes int layoutResID) {
    super.setContentView(layoutResID);
    initViews();
    setListeners();
  }

  /**
   * Call to hide soft keyboard
   *
   * @param activity calling activity
   */
  public void hideKeyboard(Activity activity) {
    InputMethodManager imm =
        (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
    // Find the currently focused view, so we can grab the correct window token from it.
    View view = activity.getCurrentFocus();
    // If no view currently has focus, create a new one, just so we can grab a window token from it
    if (view == null) {
      view = new View(activity);
    }
    if (imm != null) {
      imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }
  }

  public void showKeyBoard(View view) {
    InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
    if (imm != null) {
      imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT);
    }
  }

  /**
   * Call to disable the given button
   *
   * @param button given button
   */
  protected void disableButton(Button button) {
    button.setEnabled(false);
  }

  /**
   * Call to enable the given button
   *
   * @param button given button
   */
  protected void enableButton(Button button) {
    button.setEnabled(true);
  }

  /** Template method to initialize views of activity */
  public void initViews(){}

  /** Template method to set listeners of view or callback */
  public void setListeners(){}

  /** To get layout resource id */
  public abstract int getLayoutResId();

  public boolean isNetworkConnected() {
    ConnectivityManager cm = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
    NetworkInfo networkInfo = null;
    if (cm != null) {
      networkInfo = cm.getActiveNetworkInfo();
    }
    return networkInfo != null && networkInfo.isConnected();
  }

  public Context getContext() {
    return mContext;
  }

  @Override
  protected void onResume() {
    super.onResume();
  }

  @Override
  protected void onPause() {
    super.onPause();
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    @MenuRes int menuId = getMenuId();
    if(menuId != 0) {
      getMenuInflater().inflate(menuId, menu);
    }
    return super.onCreateOptionsMenu(menu);
  }

  /**
   * Template method to get Menu Resource Id
   *
   * @return <code>0</code> if no menu item needed to inflate, else <code>override</code> by sub activity
   */
  public int getMenuId() {
    return 0;
  }
}

File: app/src/main/java/com/fazemeright/chatbotmetcs622/ui/base/BaseActivity.kt
package com.fazemeright.chatbotmetcs622.ui.base

import android.app.Activity
import android.content.Context
import android.net.ConnectivityManager
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import androidx.annotation.MenuRes
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.viewbinding.ViewBinding

abstract class BaseActivity<VM : BaseViewModel, VB : ViewBinding> : AppCompatActivity() {

    lateinit var context: Context

    protected lateinit var binding: VB
    protected lateinit var viewModel: VM
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = inflateLayoutFromBinding()
        context = this
        viewModel = ViewModelProvider(this).get(viewModelClass::class.java)
        setContentView(binding.root)
        initViews()
        setListeners()
    }

    abstract fun inflateLayoutFromBinding(): VB

    protected abstract val viewModelClass: VM

    /**
     * Call to hide soft keyboard.
     *
     * @param activity calling activity
     */
    fun hideKeyboard(activity: Activity) {
        val imm = activity.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        // Find the currently focused view, so we can grab the correct window token from it.
        var view = activity.currentFocus
        // If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = View(activity)
        }
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    /**
     * Call to show keyboard.
     *
     * @param view view
     */
    fun showKeyBoard(view: View?) {
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT)
    }

    /**
     * Call to disable the given button.
     *
     * @param button given button
     */
    protected fun disableButton(button: Button) {
        button.isEnabled = false
    }

    /**
     * Call to enable the given button.
     *
     * @param button given button
     */
    protected fun enableButton(button: Button) {
        button.isEnabled = true
    }

    /**
     * Template method to initialize views of activity.
     */
    open fun initViews() {}

    /**
     * Template method to set listeners of view or callback.
     */
    open fun setListeners() {}

    /**
     * Determine if network is available and connected.
     *
     * @return `true` if connected, else `false`
     */
    val isNetworkConnected: Boolean
        get() {
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            return cm.isDefaultNetworkActive
        }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        @MenuRes val menuId = menuId
        if (menuId != 0) {
            menuInflater.inflate(menuId, menu)
        }
        return super.onCreateOptionsMenu(menu)
    }

    /**
     * Template method to get Menu Resource Id.
     *
     * @return `0` if no menu item needed to inflate, else `override`.
     */
    open val menuId: Int
        get() = 0
}